R/tokenizers.R

```diff
@@ -0,0 +1,170 @@
+#' @keywords tokens internal
+#' @examples
+#' \dontrun{
+#' txt <- c(doc1 = "Tweet https://quanteda.io using @quantedainit and #rstats.",
+#'          doc2 = "The £1,000,000 question.",
+#'          doc3 = "毎日 #quanteda を使用してください！",
+#'          doc4 = "Line 1.\nLine2\n\nLine3.",
+#'          doc5 = "?",
+#'          doc6 = "Self-aware machines! \U0001f600")
+#' tokenize_word(txt)
+#' tokenize_word(txt, split_hyphens = TRUE)
+#' tokenize_word1(txt, split_hyphens = FALSE)
+#' tokenize_word1(txt, split_hyphens = TRUE)
+#' tokenize_fasterword(txt)
+#' tokenize_fastestword(txt)
+#' tokenize_sentence(txt)
+#' tokenize_character(txt[2])
+#' }
+NULL
+
+# improved tokenizer ----------
+
+#' @rdname tokenize_internal
+#' @inheritParams tokens
+#' @param version integer; if 1, behave the same as pre-v2 "word" (meaning
+#'   that URLs and email addresses are split up)
+#' @importFrom stringi stri_split_regex stri_detect_regex stri_detect_charclass
+#'   stri_replace_all_regex stri_detect_fixed stri_replace_all_fixed
+#' @export
+tokenize_word <- function(x,
+                          split_tags = FALSE,
+                          split_hyphens = FALSE,
+                          version = 2,
+                          verbose = quanteda_options("verbose")) {
+
+    named <- names(x)
+
+    # -------- special processing
+    # this will not be needed if we can modify the ICU type rules to protect them
+    # remove variant selector & whitespace with diacritical marks
+    x <- stri_replace_all_regex(x, c("[\uFE00-\uFE0F]", "\\s[\u0300-\u036F]"), "",
+                                vectorize_all = FALSE)
+    # substitute characters not to split
+    x <- preserve_special(x, split_hyphens, split_tags, verbose)
+
+    # -------- initial split on white space
+    # this and the index_url can be removed if we are able to modify the rules in
+    # type for stri_split_boundaries()
+    if (verbose) catm("...segmenting tokens\n")
+    # make NAs ""
+    x[is.na(x)] <- ""
+    if (version >= 2) {
+        x <- stri_split_regex(x, "[\\p{Z}\\p{C}]+", omit_empty = FALSE)
+    }
+
+    # -------- get document indexes and then unlist tokens
+    if (verbose) catm("...vectorizing tokens\n")
+    doc_lengths <- cumsum(lengths(x))
+    docindex <- c(0, doc_lengths)
+    x <- unlist(x)
+
+    # -------- identify URLs to protext in second stage tokenization
+    index_url <- if (version >= 2) stri_detect_fixed(x, "://") else rep(FALSE, length(x))
+
+    # -------- apply ICU word tokenizer to everything not indexed for preservation
+    if (verbose) catm("...applying second stage smart tokenization\n")
+    x[!index_url] <- stri_split_boundaries(x[!index_url], type = "word")
+
+    # -------- convert the vector back to a list
+    if (verbose) catm("...tidying up\n")
+    x <- split(x, cut(seq_along(x), docindex, include.lowest = FALSE, labels = named))
+    # handle nested lists from 2nd stage stri_split_boundaries() call
+    x <- lapply(x, unlist)
```

Comment:
I don't think you need these slow  commands for word1 tokenization.

Reply:
Fixed in https://github.com/quanteda/quanteda/pull/1870